File: clients/Cpp/model/DebugData.cpp
#include "DebugData.hpp"
#include <stdexcept>

DebugData::Log::Log() { }
DebugData::Log::Log(std::string text) : text(text) { }
DebugData::Log DebugData::Log::readFrom(InputStream& stream) {
    DebugData::Log result;
    result.text = stream.readString();
    return result;
}
void DebugData::Log::writeTo(OutputStream& stream) const {
    stream.write(TAG);
    stream.write(text);
}

DebugData::Primitives::Primitives() { }
DebugData::Primitives::Primitives(std::vector<ColoredVertex> vertices, PrimitiveType primitiveType) : vertices(vertices), primitiveType(primitiveType) { }
DebugData::Primitives DebugData::Primitives::readFrom(InputStream& stream) {
    DebugData::Primitives result;
    result.vertices = std::vector<ColoredVertex>(stream.readInt());
    for (size_t i = 0; i < result.vertices.size(); i++) {
        result.vertices[i] = ColoredVertex::readFrom(stream);
    }
    switch (stream.readInt()) {
    case 0:
        result.primitiveType = PrimitiveType::LINES;
        break;
    case 1:
        result.primitiveType = PrimitiveType::TRIANGLES;
        break;
    default:
        throw std::runtime_error("Unexpected tag value");
    }
    return result;
}
void DebugData::Primitives::writeTo(OutputStream& stream) const {
    stream.write(TAG);
    stream.write((int)(vertices.size()));
    for (const ColoredVertex& verticesElement : vertices) {
        verticesElement.writeTo(stream);
    }
    stream.write((int)(primitiveType));
}

DebugData::PlacedText::PlacedText() { }
DebugData::PlacedText::PlacedText(ColoredVertex vertex, std::string text, float alignment, float size) : vertex(vertex), text(text), alignment(alignment), size(size) { }
DebugData::PlacedText DebugData::PlacedText::readFrom(InputStream& stream) {
    DebugData::PlacedText result;
    result.vertex = ColoredVertex::readFrom(stream);
    result.text = stream.readString();
    result.alignment = stream.readFloat();
    result.size = stream.readFloat();
    return result;
}
void DebugData::PlacedText::writeTo(OutputStream& stream) const {
    stream.write(TAG);
    vertex.writeTo(stream);
    stream.write(text);
    stream.write(alignment);
    stream.write(size);
}
std::shared_ptr<DebugData> DebugData::readFrom(InputStream& stream) {
    switch (stream.readInt()) {
    case 0:
        return std::shared_ptr<DebugData::Log>(new DebugData::Log(DebugData::Log::readFrom(stream)));
    case 1:
        return std::shared_ptr<DebugData::Primitives>(new DebugData::Primitives(DebugData::Primitives::readFrom(stream)));
    case 2:
        return std::shared_ptr<DebugData::PlacedText>(new DebugData::PlacedText(DebugData::PlacedText::readFrom(stream)));
    default:
        throw std::runtime_error("Unexpected tag value");
    }
};

File: clients/Go/model/debug_data.go
package model

import "io"
import . "aicup2020/stream"

type DebugData interface {
    Write(writer io.Writer)
}
func ReadDebugData(reader io.Reader) DebugData {
    switch ReadInt32(reader) {
        case 0:
            return ReadDebugDataLog(reader)
        case 1:
            return ReadDebugDataPrimitives(reader)
        case 2:
            return ReadDebugDataPlacedText(reader)
    }
    panic("Unexpected tag value")
}

type DebugDataLog struct {
    Text string
}
func NewDebugDataLog(text string) DebugDataLog {
    return DebugDataLog {
        Text: text,
    }
}
func ReadDebugDataLog(reader io.Reader) DebugDataLog {
    result := DebugDataLog {}
    result.Text = ReadString(reader)
    return result
}
func (value DebugDataLog) Write(writer io.Writer) {
    WriteInt32(writer, 0)
    WriteString(writer, value.Text)
}

type DebugDataPrimitives struct {
    Vertices []ColoredVertex
    PrimitiveType PrimitiveType
}
func NewDebugDataPrimitives(vertices []ColoredVertex, primitiveType PrimitiveType) DebugDataPrimitives {
    return DebugDataPrimitives {
        Vertices: vertices,
        PrimitiveType: primitiveType,
    }
}
func ReadDebugDataPrimitives(reader io.Reader) DebugDataPrimitives {
    result := DebugDataPrimitives {}
    result.Vertices = make([]ColoredVertex, ReadInt32(reader))
    for i := range result.Vertices {
        result.Vertices[i] = ReadColoredVertex(reader)
    }
    result.PrimitiveType = ReadPrimitiveType(reader)
    return result
}
func (value DebugDataPrimitives) Write(writer io.Writer) {
    WriteInt32(writer, 1)
    WriteInt32(writer, int32(len(value.Vertices)))
    for _, VerticesElement := range value.Vertices {
        VerticesElement.Write(writer)
    }
    WriteInt32(writer, int32(value.PrimitiveType))
}

type DebugDataPlacedText struct {
    Vertex ColoredVertex
    Text string
    Alignment float32
    Size float32
}
func NewDebugDataPlacedText(vertex ColoredVertex, text string, alignment float32, size float32) DebugDataPlacedText {
    return DebugDataPlacedText {
        Vertex: vertex,
        Text: text,
        Alignment: alignment,
        Size: size,
    }
}
func ReadDebugDataPlacedText(reader io.Reader) DebugDataPlacedText {
    result := DebugDataPlacedText {}
    result.Vertex = ReadColoredVertex(reader)
    result.Text = ReadString(reader)
    result.Alignment = ReadFloat32(reader)
    result.Size = ReadFloat32(reader)
    return result
}
func (value DebugDataPlacedText) Write(writer io.Writer) {
    WriteInt32(writer, 2)
    value.Vertex.Write(writer)
    WriteString(writer, value.Text)
    WriteFloat32(writer, value.Alignment)
    WriteFloat32(writer, value.Size)
}
